### ISSUE TITLE
DateTime parse error on non en-US system and PowerShell Core

### ISSUE THREAD

**Issue by hbuckle** [rmbolger/Posh-ACME#117]
On my en-GB system when running Submit-Renewal I get the following error

```
Get-PAOrder | Submit-Renewal
Exception calling "Parse" with "1" argument(s): "String '01/19/2019 05:38:13' was not recognized as a valid DateTime."
At C:\Program Files\PowerShell\Modules\Posh-Acme\3.1.1\Public\Submit-Renewal.ps1:52 char:21
```

There seems to be some weirdness around how PowerShell is casting the DateTime object to a string, if you explicitly call `.ToString()` then it applies the correct culture

```
$now = Get-Date
[DateTimeOffset]::Parse($now)
Exception calling "Parse" with "1" argument(s): "String '12/24/2018 07:00:32' was not recognized as a valid DateTime."
[DateTimeOffset]::Parse($now.ToString())


DateTime      : 24/12/2018 07:00:32
UtcDateTime   : 24/12/2018 07:00:32
LocalDateTime : 24/12/2018 07:00:32
Date          : 24/12/2018 00:00:00
Day           : 24
DayOfWeek     : Monday
DayOfYear     : 358
Hour          : 7
Millisecond   : 0
Minute        : 0
Month         : 12
Offset        : 00:00:00
Second        : 32
Ticks         : 636812316320000000
UtcTicks      : 636812316320000000
TimeOfDay     : 07:00:32
Year          : 2018
```

Or using the datetimeoffset constructur works as well

```
[DateTimeOffset]::new($now)
DateTime      : 24/12/2018 07:03:49
UtcDateTime   : 24/12/2018 07:03:49
LocalDateTime : 24/12/2018 07:03:49
Date          : 24/12/2018 00:00:00
Day           : 24
DayOfWeek     : Monday
DayOfYear     : 358
Hour          : 7
Millisecond   : 583
Minute        : 3
Month         : 12
Offset        : 00:00:00
Second        : 49
Ticks         : 636812318295838428
UtcTicks      : 636812318295838428
TimeOfDay     : 07:03:49.5838428
Year          : 2018
```

**Comment by hbuckle** [comment]
Forcing a new certificate order seems to have resolved the issue. Has something changed with how RenewAfter is stored? Prior to forcing the new certificate it was a DateTime

```
(Get-PAOrder).RenewAfter.GetType()

IsPublic IsSerial Name                                     BaseType
-------- -------- ----                                     --------
True     True     DateTime                                 System.ValueType
```

After the new certificate it's now a string

```
(Get-PAOrder).RenewAfter.GetType()

IsPublic IsSerial Name                                     BaseType
-------- -------- ----                                     --------
True     True     String                                   System.Object
```

**Comment by rmbolger** [comment]
As far as I can tell, RenewAfter has been an ISO 8601 string since at least v1.0. This is what it looked like at the v1.0 tag's commit.

https://github.com/rmbolger/Posh-ACME/blob/v1.0/Posh-ACME/Public/New-PACertificate.ps1#L129-L131

So that's really weird unless you were using a version prior to the 1.0 release. I didn't check any further back. Glad to hear it's working now though.

**Comment by rmbolger** [comment]
That's odd. On my system, it's still a string even after ConvertFrom-Json. The code inside `Get-PAOrder` is not much different either. It basically does this before returning the order object.

```powershell
$order = Get-ChildItem <path to order.json> | Get-Content -Raw | ConvertFrom-Json
$order.PSObject.TypeNames.Insert(0,'PoshACME.PAOrder')
```

The piece that inserts the type name just makes it so that the custom display formatter will pick it up and that it will be recognized by other functions that accept the custom type.

Do you have any other modules installed that might be adding extension methods or modifying default functionality? I seem to recall another user having weird issues with a module that was adding extension methods to the .NET String object.

**Comment by hbuckle** [comment]
OK so this is a [PowerShell 6 thing](https://github.com/PowerShell/PowerShell/issues/3378#issuecomment-296336419), the conversion doesn't happen in 5.1

Unfortunately the issue wasn't actually resolved by the new certificate either - it was just that the PAOrder was cached in the `script:` variable so it wasn't re-read from the JSON file. When I opened a new PowerShell 6 window the same problem appears again

```powershell
PS C:\> $PSVersionTable

Name                           Value
----                           -----
PSVersion                      6.1.1
PSEdition                      Core
GitCommitId                    6.1.1
OS                             Microsoft Windows 10.0.17763
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0


PS C:\> $order = Get-PAOrder
PS C:\> $order.RenewAfter.GetType()

IsPublic IsSerial Name                                     BaseType
-------- -------- ----                                     --------
True     True     DateTime                                 System.ValueType


PS C:\> $order | Submit-Renewal
Exception calling "Parse" with "1" argument(s): "String '02/28/2019 14:01:21' was not recognized as a valid DateTime."
At C:\Program Files\PowerShell\Modules\Posh-ACME\3.1.1\Public\Submit-Renewal.ps1:52 char:21
+ ...         if (!$Force -and $null -ne $order.RenewAfter -and (Get-DateTi ...
+                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ CategoryInfo          : NotSpecified: (:) [], MethodInvocationException
+ FullyQualifiedErrorId : FormatException

Exception calling "Parse" with "1" argument(s): "String '02/28/2019 14:01:21' was not recognized as a valid DateTime."
At C:\Program Files\PowerShell\Modules\Posh-ACME\3.1.1\Public\New-PACertificate.ps1:88 char:9
+     if ($Force -or !$order -or
+         ~~~~~~~~~~~~~~~~~~~~~~
+ CategoryInfo          : NotSpecified: (:) [], MethodInvocationException
+ FullyQualifiedErrorId : FormatException
```

**Comment by rmbolger** [comment]
Oh man, you nailed it. I had no idea the JSON parsing was different. That's...super annoying. Though totally my fault for not fully testing the renewal scenario under Core.

This may take a bit to deal with as there's a lot of JSON parsing all over the place that I may need to mess with. Thanks for the find!

**Comment by hbuckle** [comment]
That is indeed a very annoying change.

I did some digging and I was able to get the old behaviour back using [JsonConvert.DefaultSettings](https://www.newtonsoft.com/json/help/html/P_Newtonsoft_Json_JsonConvert_DefaultSettings.htm)

```powershell
@{date = [datetime]::Now} | ConvertTo-Json | Set-Content C:\Temp\date.json -Encoding utf8NoBOM
Get-Content C:\Temp\date.json | ConvertFrom-Json

# date is returned as a [DateTime]
date
----
31/12/2018 10:43:00


$settings = [Newtonsoft.Json.JsonSerializerSettings]::new()
$settings.DateParseHandling = "None"
[Newtonsoft.Json.JsonConvert]::DefaultSettings = [Func[Newtonsoft.Json.JsonSerializerSettings]]{$settings}

Get-Content C:\Temp\date.json | ConvertFrom-Json

# date is returned as a [String]
date
----
2018-12-31T10:43:00.8678225+00:00
```

**Comment by rmbolger** [comment]
Setting DateParseHandling to None would end up fixing it too much the other direction I think. PowerShell 5.1 will still parse the `"\/Date(1198908717056)\/"` style dates to a DateTime but leave the ISO 8601 dates alone. I'll poke around the Newtonsoft docs to see if there's a way to make it *only* ignore the ISO 8601 dates. But otherwise, I may end up just un-parsing the dates I care about when they're converted originally.

**Comment by rmbolger** [comment]
Ok, I think I've got most of the fixes committed that should squash this bug. Care to give it a whirl?

**Comment by rmbolger** [comment]
This is now fixed in the 3.2.0 release.
